File: java/l2server/gameserver/scripting/L2ScriptEngineManager.java
/*
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program. If not, see <http://www.gnu.org/licenses/>.
 */

package l2server.gameserver.scripting;

import com.l2jserver.script.jython.JythonScriptEngine;
import l2server.Config;
import l2server.log.Log;

import javax.script.*;
import java.io.*;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.logging.Level;

/**
 * Caches script engines and provides funcionality for executing and managing scripts.<BR>
 *
 * @author KenM
 */
public final class L2ScriptEngineManager
{
	public static final File SCRIPT_FOLDER =
			new File(Config.DATAPACK_ROOT.getAbsolutePath(), Config.DATA_FOLDER + "scripts");

	public static L2ScriptEngineManager getInstance()
	{
		return SingletonHolder._instance;
	}

	private final Map<String, ScriptEngine> _nameEngines = new HashMap<>();
	private final Map<String, ScriptEngine> _extEngines = new HashMap<>();
	private final List<ScriptManager<?>> _scriptManagers = new LinkedList<>();

	private final CompiledScriptCache _cache;

	private File _currentLoadingScript;

	// Configs
	// TODO move to config file
	/**
	 * Informs(logs) the scripts being loaded.<BR>
	 * Apply only when executing script from files.<BR>
	 */
	private final boolean VERBOSE_LOADING = false;

	/**
	 * If the script engine supports compilation the script is compiled before execution.<BR>
	 */
	private final boolean ATTEMPT_COMPILATION = true;

	/**
	 * Use Compiled Scripts Cache.<BR>
	 * Only works if ATTEMPT_COMPILATION is true.<BR>
	 * DISABLED DUE TO ISSUES (if a superclass file changes subclasses are not recompiled while they should)
	 */
	private final boolean USE_COMPILED_CACHE = true;

	/**
	 * Clean an previous error log(if such exists) for the script being loaded before trying to load.<BR>
	 * Apply only when executing script from files.<BR>
	 */
	private final boolean PURGE_ERROR_LOG = true;

	private L2ScriptEngineManager()
	{
		ScriptEngineManager scriptEngineManager = new ScriptEngineManager();
		List<ScriptEngineFactory> factories = scriptEngineManager.getEngineFactories();
		if (USE_COMPILED_CACHE)
		{
			_cache = loadCompiledScriptCache();
		}
		else
		{
			_cache = null;
		}
		Log.info("Initializing Script Engine Manager");

		for (ScriptEngineFactory factory : factories)
		{
			try
			{
				ScriptEngine engine = factory.getScriptEngine();
				boolean reg = false;
				for (String name : factory.getNames())
				{
					ScriptEngine existentEngine = _nameEngines.get(name);

					if (existentEngine != null)
					{
						double engineVer = Double.parseDouble(factory.getEngineVersion());
						double existentEngVer = Double.parseDouble(existentEngine.getFactory().getEngineVersion());

						if (engineVer <= existentEngVer)
						{
							continue;
						}
					}

					reg = true;
					_nameEngines.put(name, engine);
				}

				if (reg)
				{
					Log.info("Script Engine: " + factory.getEngineName() + " " + factory.getEngineVersion() +
							" - Language: " + factory.getLanguageName() + " - Language Version: " +
							factory.getLanguageVersion());
				}

				for (String ext : factory.getExtensions())
				{
					if (!ext.equals("java") || factory.getLanguageName().equals("java"))
					{
						_extEngines.put(ext, engine);
					}
				}
			}
			catch (Exception e)
			{
				Log.log(Level.WARNING, "Failed initializing factory: " + e.getMessage(), e);
			}
		}

		preConfigure();
	}

	private void preConfigure()
	{
		// java class path

		// Jython sys.path
		String dataPackDirForwardSlashes = SCRIPT_FOLDER.getPath().replaceAll("\\\\", "/");
		String configScript = "import sys;sys.path.insert(0,'" + dataPackDirForwardSlashes + "');";
		try
		{
			this.eval("jython", configScript);
		}
		catch (ScriptException e)
		{
			Log.severe("Failed preconfiguring jython: " + e.getMessage());
		}
	}

	private ScriptEngine getEngineByName(String name)
	{
		return _nameEngines.get(name);
	}

	private ScriptEngine getEngineByExtension(String ext)
	{
		return _extEngines.get(ext);
	}

	public void executeScriptList(File list) throws IOException
	{
		File file;

		if (!Config.ALT_DEV_NO_HANDLERS && Config.ALT_DEV_NO_QUESTS)
		{
			file = new File(SCRIPT_FOLDER, "handlers/MasterHandler.java");

			try
			{
				executeScript(file);
				Log.info("Handlers loaded, all other scripts skipped");
				return;
			}
			catch (ScriptException se)
			{
				Log.log(Level.WARNING, "", se);
			}
		}

		if (Config.ALT_DEV_NO_QUESTS)
		{
			return;
		}

		if (list.isFile())
		{
			LineNumberReader lnr = new LineNumberReader(new InputStreamReader(new FileInputStream(list)));
			String line;

			while ((line = lnr.readLine()) != null)
			{
				if (Config.ALT_DEV_NO_HANDLERS && line.contains("MasterHandler.java"))
				{
					continue;
				}

				String[] parts = line.trim().split("#");

				if (parts.length > 0 && !parts[0].startsWith("#") && parts[0].length() > 0)
				{
					line = parts[0];

					if (line.endsWith("/**"))
					{
						line = line.substring(0, line.length() - 3);
					}
					else if (line.endsWith("/*"))
					{
						line = line.substring(0, line.length() - 2);
					}

					file = new File(SCRIPT_FOLDER, line);

					if (file.isDirectory() && parts[0].endsWith("/**"))
					{
						executeAllScriptsInDirectory(file, true, 32);
					}
					else if (file.isDirectory() && parts[0].endsWith("/*"))
					{
						executeAllScriptsInDirectory(file);
					}
					else if (file.isFile())
					{
						try
						{
							executeScript(file);
						}
						catch (ScriptException e)
						{
							reportScriptFileError(file, e);
						}
					}
					else
					{
						Log.warning("Failed loading: (" + file.getCanonicalPath() + ") @ " + list.getName() + ":" +
								lnr.getLineNumber() + " - Reason: doesnt exists or is not a file.");
					}
				}
			}
			lnr.close();
		}
		else
		{
			throw new IllegalArgumentException("Argument must be an file containing a list of scripts to be loaded");
		}
	}

	public void executeAllScriptsInDirectory(File dir)
	{
		this.executeAllScriptsInDirectory(dir, false, 0);
	}

	public void executeAllScriptsInDirectory(File dir, boolean recurseDown, int maxDepth)
	{
		this.executeAllScriptsInDirectory(dir, recurseDown, maxDepth, 0);
	}

	private void executeAllScriptsInDirectory(File dir, boolean recurseDown, int maxDepth, int currentDepth)
	{
		if (dir.isDirectory())
		{
			for (File file : dir.listFiles())
			{
				if (file.isDirectory() && recurseDown && maxDepth > currentDepth)
				{
					if (VERBOSE_LOADING)
					{
						Log.info("Entering folder: " + file.getName());
					}
					executeAllScriptsInDirectory(file, recurseDown, maxDepth, currentDepth + 1);
				}
				else if (file.isFile())
				{
					try
					{
						String name = file.getName();
						int lastIndex = name.lastIndexOf('.');
						String extension;
						if (lastIndex != -1)
						{
							extension = name.substring(lastIndex + 1);
							ScriptEngine engine = getEngineByExtension(extension);
							if (engine != null)
							{
								executeScript(engine, file);
							}
						}
					}
					catch (FileNotFoundException e)
					{
						// should never happen
						Log.log(Level.WARNING, "", e);
					}
					catch (ScriptException e)
					{
						reportScriptFileError(file, e);
						//Logozo.log(Level.WARNING, "", e);
					}
				}
			}
		}
		else
		{
			throw new IllegalArgumentException("The argument directory either doesnt exists or is not an directory.");
		}
	}

	public CompiledScriptCache getCompiledScriptCache()
	{
		return _cache;
	}

	public CompiledScriptCache loadCompiledScriptCache()
	{
		if (USE_COMPILED_CACHE)
		{
			File file = new File(SCRIPT_FOLDER, "CompiledScripts.cache");
			if (file.isFile())
			{
				ObjectInputStream ois = null;
				try
				{
					ois = new ObjectInputStream(new FileInputStream(file));
					CompiledScriptCache cache = (CompiledScriptCache) ois.readObject();
					cache.checkFiles();
					return cache;
				}
				catch (InvalidClassException e)
				{
					Log.log(Level.SEVERE, "Failed loading Compiled Scripts Cache, invalid class (Possibly outdated).",
							e);
				}
				catch (IOException e)
				{
					Log.log(Level.SEVERE, "Failed loading Compiled Scripts Cache from file.", e);
				}
				catch (ClassNotFoundException e)
				{
					Log.log(Level.SEVERE, "Failed loading Compiled Scripts Cache, class not found.", e);
				}
				finally
				{
					try
					{
						ois.close();
					}
					catch (Exception e)
					{
						e.printStackTrace();
					}
				}
				return new CompiledScriptCache();
			}
			else
			{
				return new CompiledScriptCache();
			}
		}
		return null;
	}

	public void executeScript(File file) throws ScriptException, FileNotFoundException
	{
		String name = file.getName();
		int lastIndex = name.lastIndexOf('.');
		String extension;
		if (lastIndex != -1)
		{
			extension = name.substring(lastIndex + 1);
		}
		else
		{
			throw new ScriptException(
					"Script file (" + name + ") doesnt has an extension that identifies the ScriptEngine to be used.");
		}

		ScriptEngine engine = getEngineByExtension(extension);
		if (engine == null)
		{
			throw new ScriptException("No engine registered for extension (" + extension + ")");
		}
		else
		{
			executeScript(engine, file);
		}
	}

	public void executeScript(String engineName, File file) throws FileNotFoundException, ScriptException
	{
		ScriptEngine engine = getEngineByName(engineName);
		if (engine == null)
		{
			throw new ScriptException("No engine registered with name (" + engineName + ")");
		}
		else
		{
			executeScript(engine, file);
		}
	}

	public void executeScript(ScriptEngine engine, File file) throws FileNotFoundException, ScriptException
	{
		BufferedReader reader = new BufferedReader(new InputStreamReader(new FileInputStream(file)));

		if (VERBOSE_LOADING)
		{
			Log.info("Loading Script: " + file.getAbsolutePath());
		}

		if (PURGE_ERROR_LOG)
		{
			String name = file.getAbsolutePath() + ".error.log";
			File errorLog = new File(name);
			if (errorLog.isFile())
			{
				errorLog.delete();
			}
		}

		if (engine instanceof Compilable && ATTEMPT_COMPILATION)
		{
			ScriptContext context = new SimpleScriptContext();
			context.setAttribute("mainClass", getClassForFile(file).replace('/', '.').replace('\\', '.'),
					ScriptContext.ENGINE_SCOPE);
			context.setAttribute(ScriptEngine.FILENAME, file.getName(), ScriptContext.ENGINE_SCOPE);
			context.setAttribute("classpath", SCRIPT_FOLDER.getAbsolutePath(), ScriptContext.ENGINE_SCOPE);
			context.setAttribute("sourcepath", SCRIPT_FOLDER.getAbsolutePath(), ScriptContext.ENGINE_SCOPE);
			context.setAttribute(JythonScriptEngine.JYTHON_ENGINE_INSTANCE, engine, ScriptContext.ENGINE_SCOPE);

			setCurrentLoadingScript(file);
			ScriptContext ctx = engine.getContext();
			try
			{
				engine.setContext(context);
				if (USE_COMPILED_CACHE)
				{
					CompiledScript cs = _cache.loadCompiledScript(engine, file);
					cs.eval(context);
				}
				else
				{
					Compilable eng = (Compilable) engine;
					CompiledScript cs = eng.compile(reader);
					cs.eval(context);
				}
			}
			finally
			{
				engine.setContext(ctx);
				setCurrentLoadingScript(null);
				context.removeAttribute(ScriptEngine.FILENAME, ScriptContext.ENGINE_SCOPE);
				context.removeAttribute("mainClass", ScriptContext.ENGINE_SCOPE);
			}
		}
		else
		{
			ScriptContext context = new SimpleScriptContext();
			context.setAttribute("mainClass", getClassForFile(file).replace('/', '.').replace('\\', '.'),
					ScriptContext.ENGINE_SCOPE);
			context.setAttribute(ScriptEngine.FILENAME, file.getName(), ScriptContext.ENGINE_SCOPE);
			context.setAttribute("classpath", SCRIPT_FOLDER.getAbsolutePath(), ScriptContext.ENGINE_SCOPE);
			context.setAttribute("sourcepath", SCRIPT_FOLDER.getAbsolutePath(), ScriptContext.ENGINE_SCOPE);
			setCurrentLoadingScript(file);
			try
			{
				engine.eval(reader, context);
			}
			finally
			{
				setCurrentLoadingScript(null);
				engine.getContext().removeAttribute(ScriptEngine.FILENAME, ScriptContext.ENGINE_SCOPE);
				engine.getContext().removeAttribute("mainClass", ScriptContext.ENGINE_SCOPE);
			}
		}
	}

	public static String getClassForFile(File script)
	{
		String path = script.getAbsolutePath();
		String scpPath = SCRIPT_FOLDER.getAbsolutePath();
		if (path.startsWith(scpPath))
		{
			int idx = path.lastIndexOf('.');
			return path.substring(scpPath.length() + 1, idx);
		}
		return null;
	}

	public ScriptContext getScriptContext(ScriptEngine engine)
	{
		return engine.getContext();
	}

	public ScriptContext getScriptContext(String engineName)
	{
		ScriptEngine engine = getEngineByName(engineName);
		if (engine == null)
		{
			throw new IllegalStateException("No engine registered with name (" + engineName + ")");
		}
		else
		{
			return this.getScriptContext(engine);
		}
	}

	public Object eval(ScriptEngine engine, String script, ScriptContext context) throws ScriptException
	{
		if (engine instanceof Compilable && ATTEMPT_COMPILATION)
		{
			Compilable eng = (Compilable) engine;
			CompiledScript cs = eng.compile(script);
			return context != null ? cs.eval(context) : cs.eval();
		}
		else
		{
			return context != null ? engine.eval(script, context) : engine.eval(script);
		}
	}

	public Object eval(String engineName, String script) throws ScriptException
	{
		return this.eval(engineName, script, null);
	}

	public Object eval(String engineName, String script, ScriptContext context) throws ScriptException
	{
		ScriptEngine engine = getEngineByName(engineName);
		if (engine == null)
		{
			throw new ScriptException("No engine registered with name (" + engineName + ")");
		}
		else
		{
			return this.eval(engine, script, context);
		}
	}

	public Object eval(ScriptEngine engine, String script) throws ScriptException
	{
		return this.eval(engine, script, null);
	}

	public void reportScriptFileError(File script, ScriptException e)
	{
		String dir = script.getParent();
		if (dir != null)
		{
			String errorHeader =
					"Error on: " + script.getAbsolutePath() + "\r\nLine: " + e.getLineNumber() + " - Column: " +
							e.getColumnNumber() + "\r\n\r\n";
			Log.warning("Failed executing script: " + script.getAbsolutePath() + ":");
			Log.warning(errorHeader);
			Log.warning(e.getMessage());
		}
		else
		{
			Log.log(Level.WARNING, "Failed executing script: " + script.getAbsolutePath() + "\r\n" + e.getMessage() +
					"Additionally failed when trying to write an error report on script directory.", e);
		}
	}

	public void registerScriptManager(ScriptManager<?> manager)
	{
		_scriptManagers.add(manager);
	}

	public void removeScriptManager(ScriptManager<?> manager)
	{
		_scriptManagers.remove(manager);
	}

	public List<ScriptManager<?>> getScriptManagers()
	{
		return _scriptManagers;
	}

	/**
	 * @param currentLoadingScript The currentLoadingScript to set.
	 */
	protected void setCurrentLoadingScript(File currentLoadingScript)
	{
		_currentLoadingScript = currentLoadingScript;
	}

	/**
	 * @return Returns the currentLoadingScript.
	 */
	protected File getCurrentLoadingScript()
	{
		return _currentLoadingScript;
	}

	@SuppressWarnings("synthetic-access")
	private static class SingletonHolder
	{
		protected static final L2ScriptEngineManager _instance = new L2ScriptEngineManager();
	}
}

File: src/main/java/l2server/gameserver/scripting/L2ScriptEngineManager.kt
/*
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program. If not, see <http://www.gnu.org/licenses/>.
 */

package l2server.gameserver.scripting

import com.l2jserver.script.jython.JythonScriptEngine
import l2server.Config
import l2server.gameserver.datatables.DoorTable
import l2server.gameserver.datatables.SpawnTable
import l2server.gameserver.instancemanager.QuestManager
import l2server.gameserver.instancemanager.TransformationManager
import l2server.util.loader.annotations.Load
import org.slf4j.LoggerFactory
import java.io.*
import java.util.*
import javax.script.*

/**
 * Caches script engines and provides funcionality for executing and managing scripts.<BR></BR>
 *
 * @author KenM
 */
object L2ScriptEngineManager {

    private val log = LoggerFactory.getLogger(L2ScriptEngineManager::class.java)

    @JvmStatic
    val SCRIPT_FOLDER = File(Config.DATAPACK_ROOT.absolutePath, Config.DATA_FOLDER + "scripts")

    private val nameEngines = HashMap<String, ScriptEngine>()
    private val extEngines = HashMap<String, ScriptEngine>()
    private val scriptManagers = LinkedList<ScriptManager<*>>()

    var cache: CompiledScriptCache? = null

    /**
     * @return Returns the currentLoadingScript.
     */
    /**
     * @param currentLoadingScript The currentLoadingScript to set.
     */
    var currentLoadingScript: File? = null
        private set

    // Configs
    // TODO move to config file
    /**
     * Informs(logs) the scripts being loaded.<BR></BR>
     * Apply only when executing script from files.<BR></BR>
     */
    private val VERBOSE_LOADING = false

    /**
     * If the script engine supports compilation the script is compiled before execution.<BR></BR>
     */
    private val ATTEMPT_COMPILATION = true

    /**
     * Use Compiled Scripts Cache.<BR></BR>
     * Only works if ATTEMPT_COMPILATION is true.<BR></BR>
     * DISABLED DUE TO ISSUES (if a superclass file changes subclasses are not recompiled while they should)
     */
    private val USE_COMPILED_CACHE = true

    /**
     * Clean an previous error log(if such exists) for the script being loaded before trying to load.<BR></BR>
     * Apply only when executing script from files.<BR></BR>
     */
    private val PURGE_ERROR_LOG = true

    @Load(dependencies = [SpawnTable::class, DoorTable::class])
    fun initialize() {
        val scriptEngineManager = ScriptEngineManager()
        val factories = scriptEngineManager.engineFactories
        if (USE_COMPILED_CACHE) {
            cache = loadCompiledScriptCache()
        } else {
            cache = null
        }
        log.info("Initializing Script Engine Manager")

        for (factory in factories) {
            try {
                val engine = factory.scriptEngine
                var reg = false
                for (name in factory.names) {
                    val existentEngine = nameEngines[name]

                    if (existentEngine != null) {
                        val engineVer = java.lang.Double.parseDouble(factory.engineVersion)
                        val existentEngVer = java.lang.Double.parseDouble(existentEngine.factory.engineVersion)

                        if (engineVer <= existentEngVer) {
                            continue
                        }
                    }

                    reg = true
                    nameEngines[name] = engine
                }

                if (reg) {
                    log.info("Script Engine: " + factory.engineName + " " + factory.engineVersion + " - Language: " +
                            factory.languageName + " - Language Version: " + factory.languageVersion)
                }

                for (ext in factory.extensions) {
                    if (ext != "java" || factory.languageName == "java") {
                        extEngines[ext] = engine
                    }
                }
            } catch (e: Exception) {
                log.warn("Failed initializing factory: " + e.message, e)
            }

        }

        preConfigure()

        try {
            log.info("Loading Server Scripts")
            var scripts = File(Config.DATAPACK_ROOT.toString() + "/" + Config.DATA_FOLDER + "scripts.cfg")
            if (!Config.ALT_DEV_NO_HANDLERS || !Config.ALT_DEV_NO_QUESTS) {
                L2ScriptEngineManager.executeScriptList(scripts)

                scripts = File(Config.DATAPACK_ROOT.toString() + "/data_" + Config.SERVER_NAME + "/scripts.cfg")
                if (scripts.exists()) {
                    L2ScriptEngineManager.executeScriptList(scripts)
                }
            }
        } catch (ioe: IOException) {
            log.error("Failed loading scripts.cfg, no script going to be loaded")
        }

        try {
            val compiledScriptCache = L2ScriptEngineManager.cache
            if (compiledScriptCache == null) {
                log.info("Compiled Scripts Cache is disabled.")
            } else {
                compiledScriptCache.purge()

                if (compiledScriptCache.isModified) {
                    compiledScriptCache.save()
                    log.info("Compiled Scripts Cache was saved.")
                } else {
                    log.info("Compiled Scripts Cache is up-to-date.")
                }
            }
        } catch (e: IOException) {
            log.error("Failed to store Compiled Scripts Cache.", e)
        }

        QuestManager.getInstance().report()
        TransformationManager.getInstance().report()
    }

    private fun preConfigure() {
        // java class path

        // Jython sys.path
        val dataPackDirForwardSlashes = SCRIPT_FOLDER.path.replace("\\\\".toRegex(), "/")
        val configScript = "import sys;sys.path.insert(0,'$dataPackDirForwardSlashes');"
        try {
            this.eval("jython", configScript)
        } catch (e: ScriptException) {
            log.error("Failed preconfiguring jython: " + e.message)
        }

    }

    private fun getEngineByName(name: String): ScriptEngine? {
        return nameEngines[name]
    }

    private fun getEngineByExtension(ext: String): ScriptEngine? {
        return extEngines[ext]
    }

    @Throws(IOException::class)
    fun executeScriptList(list: File) {
        var file: File

        if (!Config.ALT_DEV_NO_HANDLERS && Config.ALT_DEV_NO_QUESTS) {
            file = File(SCRIPT_FOLDER, "handlers/MasterHandler.java")

            try {
                executeScript(file)
                log.info("Handlers loaded, all other scripts skipped")
                return
            } catch (se: ScriptException) {
                log.warn("", se)
            }

        }

        if (Config.ALT_DEV_NO_QUESTS) {
            return
        }

        if (list.isFile) {
            val lnr = LineNumberReader(InputStreamReader(FileInputStream(list)))
            var line = lnr.readLine()
            while (line != null) {
                if (Config.ALT_DEV_NO_HANDLERS && line.contains("MasterHandler.java")) {
                    continue
                }

                val parts = line.trim { it <= ' ' }.split("#".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

                if (parts.size > 0 && !parts[0].startsWith("#") && parts[0].length > 0) {
                    line = parts[0]

                    if (line.endsWith("/**")) {
                        line = line.substring(0, line.length - 3)
                    } else if (line.endsWith("/*")) {
                        line = line.substring(0, line.length - 2)
                    }

                    file = File(SCRIPT_FOLDER, line)

                    if (file.isDirectory && parts[0].endsWith("/**")) {
                        executeAllScriptsInDirectory(file, true, 32)
                    } else if (file.isDirectory && parts[0].endsWith("/*")) {
                        executeAllScriptsInDirectory(file)
                    } else if (file.isFile) {
                        try {
                            executeScript(file)
                        } catch (e: ScriptException) {
                            reportScriptFileError(file, e)
                        }

                    } else {
                        log.warn("Failed loading: (" + file.canonicalPath + ") @ " + list.name + ":" + lnr.lineNumber +
                                " - Reason: it doesn't exist or it's not a file.")
                    }
                }

                line = lnr.readLine()
            }
            lnr.close()
        } else {
            throw IllegalArgumentException("Argument must be an file containing a list of scripts to be loaded")
        }
    }

    fun executeAllScriptsInDirectory(dir: File) {
        this.executeAllScriptsInDirectory(dir, false, 0)
    }

    fun executeAllScriptsInDirectory(dir: File, recurseDown: Boolean, maxDepth: Int) {
        this.executeAllScriptsInDirectory(dir, recurseDown, maxDepth, 0)
    }

    private fun executeAllScriptsInDirectory(dir: File, recurseDown: Boolean, maxDepth: Int, currentDepth: Int) {
        if (dir.isDirectory) {
            for (file in dir.listFiles()!!) {
                if (file.isDirectory && recurseDown && maxDepth > currentDepth) {
                    if (VERBOSE_LOADING) {
                        log.info("Entering folder: " + file.name)
                    }
                    executeAllScriptsInDirectory(file, recurseDown, maxDepth, currentDepth + 1)
                } else if (file.isFile) {
                    try {
                        val name = file.name
                        val lastIndex = name.lastIndexOf('.')
                        val extension: String
                        if (lastIndex != -1) {
                            extension = name.substring(lastIndex + 1)
                            val engine = getEngineByExtension(extension)
                            if (engine != null) {
                                executeScript(engine, file)
                            }
                        }
                    } catch (e: FileNotFoundException) {
                        // should never happen
                        log.warn("", e)
                    } catch (e: ScriptException) {
                        reportScriptFileError(file, e)
                        //Logozo.warn( "", e);
                    }

                }
            }
        } else {
            throw IllegalArgumentException("The argument directory either doesn't exist or it's not a directory.")
        }
    }

    fun loadCompiledScriptCache(): CompiledScriptCache? {
        if (USE_COMPILED_CACHE) {
            val file = File(SCRIPT_FOLDER, "CompiledScripts.cache")
            if (file.isFile) {
                var ois: ObjectInputStream? = null
                try {
                    ois = ObjectInputStream(FileInputStream(file))
                    val cache = ois.readObject() as CompiledScriptCache
                    cache.checkFiles()
                    return cache
                } catch (e: InvalidClassException) {
                    log.error("Failed loading Compiled Scripts Cache, invalid class (Possibly outdated).", e)
                } catch (e: IOException) {
                    log.error("Failed loading Compiled Scripts Cache from file.", e)
                } catch (e: ClassNotFoundException) {
                    log.error("Failed loading Compiled Scripts Cache, class not found.", e)
                } finally {
                    try {
                        ois!!.close()
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }

                }
                return CompiledScriptCache()
            } else {
                return CompiledScriptCache()
            }
        }
        return null
    }

    @Throws(ScriptException::class, FileNotFoundException::class)
    fun executeScript(file: File) {
        val name = file.name
        val lastIndex = name.lastIndexOf('.')
        val extension: String
        if (lastIndex != -1) {
            extension = name.substring(lastIndex + 1)
        } else {
            throw ScriptException("Script file ($name) doesnt has an extension that identifies the ScriptEngine to be used.")
        }

        val engine = getEngineByExtension(extension)
        if (engine == null) {
            throw ScriptException("No engine registered for extension ($extension)")
        } else {
            executeScript(engine, file)
        }
    }

    @Throws(FileNotFoundException::class, ScriptException::class)
    fun executeScript(engineName: String, file: File) {
        val engine = getEngineByName(engineName)
        if (engine == null) {
            throw ScriptException("No engine registered with name ($engineName)")
        } else {
            executeScript(engine, file)
        }
    }

    @Throws(FileNotFoundException::class, ScriptException::class)
    fun executeScript(engine: ScriptEngine, file: File) {
        val reader = BufferedReader(InputStreamReader(FileInputStream(file)))

        if (VERBOSE_LOADING) {
            log.info("Loading Script: " + file.absolutePath)
        }

        if (PURGE_ERROR_LOG) {
            val name = file.absolutePath + ".error.log"
            val errorLog = File(name)
            if (errorLog.isFile) {
                errorLog.delete()
            }
        }

        if (engine is Compilable && ATTEMPT_COMPILATION) {
            val context = SimpleScriptContext()
            context.setAttribute("mainClass", getClassForFile(file)!!.replace('/', '.').replace('\\', '.'), ScriptContext.ENGINE_SCOPE)
            context.setAttribute(ScriptEngine.FILENAME, file.name, ScriptContext.ENGINE_SCOPE)
            context.setAttribute("classpath", SCRIPT_FOLDER.absolutePath, ScriptContext.ENGINE_SCOPE)
            context.setAttribute("sourcepath", SCRIPT_FOLDER.absolutePath, ScriptContext.ENGINE_SCOPE)
            context.setAttribute(JythonScriptEngine.JYTHON_ENGINE_INSTANCE, engine, ScriptContext.ENGINE_SCOPE)

            currentLoadingScript = file
            val ctx = engine.context
            try {
                engine.context = context
                if (USE_COMPILED_CACHE) {
                    val cs = cache!!.loadCompiledScript(engine, file)
                    cs.eval(context)
                } else {
                    val eng = engine as Compilable
                    val cs = eng.compile(reader)
                    cs.eval(context)
                }
            } finally {
                engine.context = ctx
                currentLoadingScript = null
                context.removeAttribute(ScriptEngine.FILENAME, ScriptContext.ENGINE_SCOPE)
                context.removeAttribute("mainClass", ScriptContext.ENGINE_SCOPE)
            }
        } else {
            val context = SimpleScriptContext()
            context.setAttribute("mainClass", getClassForFile(file)!!.replace('/', '.').replace('\\', '.'), ScriptContext.ENGINE_SCOPE)
            context.setAttribute(ScriptEngine.FILENAME, file.name, ScriptContext.ENGINE_SCOPE)
            context.setAttribute("classpath", SCRIPT_FOLDER.absolutePath, ScriptContext.ENGINE_SCOPE)
            context.setAttribute("sourcepath", SCRIPT_FOLDER.absolutePath, ScriptContext.ENGINE_SCOPE)
            currentLoadingScript = file
            try {
                engine.eval(reader, context)
            } finally {
                currentLoadingScript = null
                engine.context.removeAttribute(ScriptEngine.FILENAME, ScriptContext.ENGINE_SCOPE)
                engine.context.removeAttribute("mainClass", ScriptContext.ENGINE_SCOPE)
            }
        }
    }

    fun getScriptContext(engine: ScriptEngine): ScriptContext {
        return engine.context
    }

    fun getScriptContext(engineName: String): ScriptContext {
        val engine = getEngineByName(engineName)
        return if (engine == null) {
            throw IllegalStateException("No engine registered with name ($engineName)")
        } else {
            this.getScriptContext(engine)
        }
    }

    @Throws(ScriptException::class)
    fun eval(engine: ScriptEngine, script: String, context: ScriptContext?): Any? {
        if (engine is Compilable && ATTEMPT_COMPILATION) {
            val eng = engine as Compilable
            val cs = eng.compile(script)
            return if (context != null) cs.eval(context) else cs.eval()
        } else {
            return if (context != null) engine.eval(script, context) else engine.eval(script)
        }
    }

    @Throws(ScriptException::class)
    fun eval(engineName: String, script: String): Any? {
        return this.eval(engineName, script, null)
    }

    @Throws(ScriptException::class)
    fun eval(engineName: String, script: String, context: ScriptContext?): Any? {
        val engine = getEngineByName(engineName)
        return if (engine == null) {
            throw ScriptException("No engine registered with name ($engineName)")
        } else {
            this.eval(engine, script, context)
        }
    }

    @Throws(ScriptException::class)
    fun eval(engine: ScriptEngine, script: String): Any? {
        return this.eval(engine, script, null)
    }

    fun reportScriptFileError(script: File, e: ScriptException) {
        val dir = script.parent
        if (dir != null) {
            val errorHeader = "Error on: " + script.absolutePath + "\r\nLine: " + e.lineNumber + " - Column: " + e.columnNumber + "\r\n\r\n"
            log.warn("Failed executing script: " + script.absolutePath + ":")
            log.warn(errorHeader)
            log.warn(e.message)
        } else {
            log.warn(
                    "Failed executing script: " + script.absolutePath + "\r\n" + e.message +
                            "Additionally failed when trying to write an error report on script directory.",
                    e)
        }
    }

    fun registerScriptManager(manager: ScriptManager<*>) {
        scriptManagers.add(manager)
    }

    fun removeScriptManager(manager: ScriptManager<*>) {
        scriptManagers.remove(manager)
    }

    fun getScriptManagers(): List<ScriptManager<*>> {
        return scriptManagers
    }

    fun getClassForFile(script: File): String? {
        val path = script.absolutePath
        val scpPath = SCRIPT_FOLDER.absolutePath
        if (path.startsWith(scpPath)) {
            val idx = path.lastIndexOf('.')
            return path.substring(scpPath.length + 1, idx)
        }
        return null
    }
}
